### PR TITLE
Make ScratchMemorySpace semi-regular

### PR DESCRIPTION
**Pull request by masterleinad**
**edited** Making `ScratchMemorySpace` easier to use and to reason about since could not think of any good reason not to do so.  The copy-assignment was intentionally deleted in the past but there was no comment explaining that choice.

---
~~These seem to not be defined.
Since there is another constructor deleting the private default constructor doesn't seem to be any effect. I could be convinced that the copy constructor should be deleted, though.~~

Don't use the pre-C++11 way to delete functions but use the rule of 5 instead and also allow move assignment/move construction.

### CODE REVIEW

**Comment by dalg24** [comment]
Of course they are not defined.  That's the old way to delete them.
What is the rational to revisit?

**Comment by masterleinad** [comment]
I was trying to understand the Travis problems with #3306 and was stumbling over this class.

**Comment by dalg24** [comment]
Why did you close?  The issue with your PR is not the changes you propose per se, but how you present them.
You are correct that the default constructor is implicitly deleted because we define a constructor.
But you do not discuss the changes to the assignment operator and nothing neither in the title line nor in your description highlight that change.

File: core/src/Kokkos_ScratchSpace.hpp [hunk 1]
```diff
@@ -78,9 +78,6 @@ class ScratchMemorySpace {
   mutable int m_offset;
   mutable int m_default_level;
 
-  ScratchMemorySpace();
-  ScratchMemorySpace& operator=(const ScratchMemorySpace&);
```
Comment (dhollman) [inline core/src/Kokkos_ScratchSpace.hpp:82]:
```  
  ScratchMemorySpace() = delete;
  ScratchMemorySpace& operator=(const ScratchMemorySpace&) = delete;
```

**Review by nliber — COMMENTED** [review]
This is saying we want ScratchMemorySpace to be copy constructible but not copy assignable.  Is that correct?

I'd much prefer the Rule of 5 here, which also has the advantage of not needing to explicitly delete the default constructor.

**Comment by masterleinad** [comment]
> This is saying we want ScratchMemorySpace to be copy constructible but not copy assignable. Is that correct?

Yes, we need it to be copy constructible for TeamMembers to be copy constructible.
I was trying to bring up the question if it makes sense to also make it copy assignable (since all the member functions are `const` anyway) but there is currently really no need.
Let's just keep the current behavior (and allow move assignment/move construction) following the rule of 5 as well.

**Comment by nliber** [comment]
Given that all the member variables are either `int` or `char*`, if we have copy assignment, it would do the exact same thing as move assignment.  I'd rather see that and use the rule of 0.

**Comment by masterleinad** [comment]
Since that's what I originally proposed, that's fine with me. :slightly_smiling_face:

**Comment by Rombur** [comment]
You need to apply clang-format

**Comment by masterleinad** [comment]
With the latest changes, `ScratchMemorySpace` is semiregular now.

**Review by dalg24 — APPROVED** [review]

**Review by dalg24 — APPROVED** [review]

**Pull request merged**